# DL IE 1

## Import relevent libarbries

In [186]:
import tensorflow as tf
import pandas as pd
import numpy as np
from sklearn import datasets
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

Iris Dataset:-
1. 1st column- Sepal Length
2. 2nd column- Sepal Width
3. 3rd column- Petal Length
4. 4th column- Petal Width

In [187]:
iris = datasets.load_iris()
x = iris.data
y = iris.target

print("data  + target")
for i in range (iris.data.shape[0]):
    print(x[i] ,"-> ", y[i])

data  + target
[5.1 3.5 1.4 0.2] ->  0
[4.9 3.  1.4 0.2] ->  0
[4.7 3.2 1.3 0.2] ->  0
[4.6 3.1 1.5 0.2] ->  0
[5.  3.6 1.4 0.2] ->  0
[5.4 3.9 1.7 0.4] ->  0
[4.6 3.4 1.4 0.3] ->  0
[5.  3.4 1.5 0.2] ->  0
[4.4 2.9 1.4 0.2] ->  0
[4.9 3.1 1.5 0.1] ->  0
[5.4 3.7 1.5 0.2] ->  0
[4.8 3.4 1.6 0.2] ->  0
[4.8 3.  1.4 0.1] ->  0
[4.3 3.  1.1 0.1] ->  0
[5.8 4.  1.2 0.2] ->  0
[5.7 4.4 1.5 0.4] ->  0
[5.4 3.9 1.3 0.4] ->  0
[5.1 3.5 1.4 0.3] ->  0
[5.7 3.8 1.7 0.3] ->  0
[5.1 3.8 1.5 0.3] ->  0
[5.4 3.4 1.7 0.2] ->  0
[5.1 3.7 1.5 0.4] ->  0
[4.6 3.6 1.  0.2] ->  0
[5.1 3.3 1.7 0.5] ->  0
[4.8 3.4 1.9 0.2] ->  0
[5.  3.  1.6 0.2] ->  0
[5.  3.4 1.6 0.4] ->  0
[5.2 3.5 1.5 0.2] ->  0
[5.2 3.4 1.4 0.2] ->  0
[4.7 3.2 1.6 0.2] ->  0
[4.8 3.1 1.6 0.2] ->  0
[5.4 3.4 1.5 0.4] ->  0
[5.2 4.1 1.5 0.1] ->  0
[5.5 4.2 1.4 0.2] ->  0
[4.9 3.1 1.5 0.2] ->  0
[5.  3.2 1.2 0.2] ->  0
[5.5 3.5 1.3 0.2] ->  0
[4.9 3.6 1.4 0.1] ->  0
[4.4 3.  1.3 0.2] ->  0
[5.1 3.4 1.5 0.2] ->  0
[5.  3.5 1.3 0.3] ->  0
[

Now we split this data into training and test

In [188]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

1. test_size = 0.2 shows that we are dividing the dataset into 80 - 20 pattern.
2. random_state = 42 , will shuffle the data while dividing, in same order each time we run it again

We assign variables for each hyperparameter to make fine tuning the hyperparameters easier

### Implementing Early Stopping
1. Early Stopping is used so that our model would not over fit
2. We no. of epochs are high, there is a change that our model overfits, this can be checked, when our model loss decreases but validation loss increases at some point
3. We avoid this my stopping the training, we we get our first validation loss increase
4. But when the increase in validation loss is significantly low, we let some epochs slide by.
5. This is donr by adding the 'pateince' argument in Early Stopping

### Explaination for keras.Sequential
1. We have added 2 hidden layers of size 50.
2. Activation function used - 'relu', the activation function 'sigmoid' can also be used
3. In the output layer, we have used 'softmax' activation function.
4. The softmax activation function gives us the probability of a output being in any class, this value lies between 0 and 1.

### Explaination for compile
1. We used optimizer 'adam' , it is latest optimizer widely used by professionals.
2. reason- it takes into account that our model increases the learning rate in 1st part of training and we reach the global minimum the learning rate decrease will lower.
3. also, 'adam' takes into account for 'momentum', as such our model does not reach a local minimum, it reaches to global minimum.
4. loss function used is sparse categorial crossentropy, when we are dealing with classification problem, it is always better to rely on the concept of cross entropy rather that L2 norm, etc. cross entropy is favoured because, when there is huge difference between actual and predicted values during training phase, the cross entropy gives huge value and similarly low for lower difference.
5. The argument metrics defines what we want to be displayed during training process, we have set it to display 'accuracy'.

In [189]:
input_size = 4
hidden_layer_1 = 50
hidden_layer_2 = 50
output_size = 3

early_stopping = tf.keras.callbacks.EarlyStopping(patience=7)


model = tf.keras.Sequential([
                                tf.keras.layers.Dense(hidden_layer_1 , activation= 'relu'),
                                tf.keras.layers.Dense(hidden_layer_2 , activation= 'relu'),
                                tf.keras.layers.Dense(output_size , activation='softmax')
                            ])

model.compile( optimizer = 'adam' , loss = 'sparse_categorical_crossentropy', metrics= ['accuracy'])

Epochs = 100
Batch_Size = 32
Validation_Split = 0.2 


model.fit(x_train, y_train, epochs = Epochs, batch_size=Batch_Size, validation_split=Validation_Split,callbacks=[early_stopping])


Epoch 1/100


3/3 [==============================] - 1s 185ms/step - loss: 1.1339 - accuracy: 0.3229 - val_loss: 0.8836 - val_accuracy: 0.7500
Epoch 2/100
3/3 [==============================] - 0s 19ms/step - loss: 0.9805 - accuracy: 0.5208 - val_loss: 0.8723 - val_accuracy: 0.4583
Epoch 3/100
3/3 [==============================] - 0s 23ms/step - loss: 0.9047 - accuracy: 0.3958 - val_loss: 0.8550 - val_accuracy: 0.5000
Epoch 4/100
3/3 [==============================] - 0s 23ms/step - loss: 0.8486 - accuracy: 0.7500 - val_loss: 0.8140 - val_accuracy: 0.9583
Epoch 5/100
3/3 [==============================] - 0s 23ms/step - loss: 0.8085 - accuracy: 0.9375 - val_loss: 0.7577 - val_accuracy: 0.9583
Epoch 6/100
3/3 [==============================] - 0s 22ms/step - loss: 0.7722 - accuracy: 0.8333 - val_loss: 0.7148 - val_accuracy: 0.8750
Epoch 7/100
3/3 [==============================] - 0s 23ms/step - loss: 0.7338 - accuracy: 0.8333 - val_loss: 0.6807 - val_accuracy: 0.9583
Epoch 8/100
3/3 [==============

In [190]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)

1/1 [==============================] - 0s 52ms/step - loss: 0.1284 - accuracy: 0.9667


### We Got Accuracy of 96.67 %

# Viva Questions

1. How many parameters are there in your model?
- We know that Parameters = (input_size + 1) * output_size
- Dense Layer 1:

- Input size: input_size = 4
Output size: hidden_layer_1 = 50
Parameters = (4 + 1) * 50 = 250 parameters
-Dense Layer 2:

- Input size: hidden_layer_1 = 50 (output from the previous layer)
Output size: hidden_layer_2 = 50
Parameters = (50 + 1) * 50 = 2550 parameters
- Dense Layer 3 (Output Layer):

- Input size: hidden_layer_2 = 50 (output from the previous layer)
Output size: output_size = 3.
Parameters = (50 + 1) * 3 = 153 parameters.
- Total Parameters = Parameters in Dense Layer 1 + Parameters in Dense Layer 2 + Parameters in Dense Layer 3
Total Parameters = 250 + 2550 + 153 = 2953 parameters


2. Modify your model for nonlinear classification?
- model = tf.keras.Sequential([
-                               tf.keras.layers.Dense(hidden_layer_1 , activation= 'sigmoid'),
                                tf.keras.layers.Dense(hidden_layer_2 , activation= 'sigmoid'),
                                tf.keras.layers.Dense(output_size , activation='softmax')
                            ])
- We use sigmoid activation function to bring non linearity into the model
    

    
3. What is the accuracy of your model? How can you improve it?
- Accuracy of 96.67%, We can improve this fine tuning the hyperparamters like hidden layer size, no. of hidden layers in the model, changing and experimenting various activation functions, batch size, 'pateince' argument of early stopping
